ext/MatrixAlgebraKitMooncakeExt/MatrixAlgebraKitMooncakeExt.jl

```diff
@@ -0,0 +1,352 @@
+    dS_ = Mooncake.tangent(S_dS)
+    A_ = Mooncake.primal(A_dA)
+    dA_ = Mooncake.tangent(A_dA)
+    A, dA = arrayify(A_, dA_)
+    S, dS = arrayify(S_, dS_)
+    U, nS, Vᴴ = svd_compact(A, Mooncake.primal(alg_dalg))
+    copy!(S, diagview(nS))
+    function dsvd_vals_adjoint(::Mooncake.NoRData)
+        svd_pullback!(dA, A, (U, S, Vᴴ), (nothing, dS, nothing))
+        MatrixAlgebraKit.zero!(dS)
+        return Mooncake.NoRData(), Mooncake.NoRData(), Mooncake.NoRData(), Mooncake.NoRData()
+    end
+    return S_dS, dsvd_vals_adjoint
+end
+
+@is_primitive Mooncake.DefaultCtx Mooncake.ReverseMode Tuple{typeof(MatrixAlgebraKit.svd_vals), Any, MatrixAlgebraKit.AbstractAlgorithm}
+function Mooncake.rrule!!(::CoDual{typeof(MatrixAlgebraKit.svd_vals)}, A_dA::CoDual, alg_dalg::CoDual)
+    # compute primal
+    A = Mooncake.primal(A_dA)
+    dA_ = Mooncake.tangent(A_dA)
+    A, dA = arrayify(A, dA_)
+    S = svd_vals(A, Mooncake.primal(alg_dalg))
+    U, _, Vᴴ = svd_compact(A, Mooncake.primal(alg_dalg))
+    # fdata call here is necessary to convert complicated Tangent type (e.g. of a Diagonal
+    # of ComplexF32) into the correct **forwards** data type (since we are now in the forward
+    # pass). For many types this is done automatically when the forward step returns, but
+    # not for nested structs with various fields (like Diagonal{Complex})
+    S_codual = Mooncake.CoDual(S, Mooncake.fdata(Mooncake.zero_tangent(S)))
+    function dsvd_vals_adjoint(::Mooncake.NoRData)
+        S = Mooncake.primal(S_codual)
+        dS_ = Mooncake.tangent(S_codual)
+        S, dS = Mooncake.arrayify(S, dS_)
+        svd_pullback!(dA, A, (U, S, Vᴴ), (nothing, dS, nothing))
+        MatrixAlgebraKit.zero!(dS)
+        return Mooncake.NoRData(), Mooncake.NoRData(), Mooncake.NoRData()
+    end
+    return S_codual, dsvd_vals_adjoint
+end
+
+@is_primitive Mooncake.DefaultCtx Mooncake.ReverseMode Tuple{typeof(MatrixAlgebraKit.svd_trunc), Any, MatrixAlgebraKit.TruncatedAlgorithm}
+function Mooncake.rrule!!(::CoDual{typeof(MatrixAlgebraKit.svd_trunc)}, A_dA::CoDual, alg_dalg::CoDual)
+    # compute primal
+    A_ = Mooncake.primal(A_dA)
+    dA_ = Mooncake.tangent(A_dA)
+    A, dA = arrayify(A_, dA_)
+    alg = Mooncake.primal(alg_dalg)
+    output = svd_trunc(A, alg)
+    # fdata call here is necessary to convert complicated Tangent type (e.g. of a Diagonal
+    # of ComplexF32) into the correct **forwards** data type (since we are now in the forward
+    # pass). For many types this is done automatically when the forward step returns, but
+    # not for nested structs with various fields (like Diagonal{Complex})
+    output_codual = Mooncake.CoDual(output, Mooncake.fdata(Mooncake.zero_tangent(output)))
+    function dsvd_trunc_adjoint(dy::Tuple{Mooncake.NoRData, Mooncake.NoRData, Mooncake.NoRData, T}) where {T <: Real}
+        Utrunc, Strunc, Vᴴtrunc, ϵ = Mooncake.primal(output_codual)
+        dUtrunc_, dStrunc_, dVᴴtrunc_, dϵ = Mooncake.tangent(output_codual)
```

Comment:
We currently do not support a nontrivial cotangent for the truncation error, but this could be done (and is probably not even that hard). However, how should I within Mooncake world now read of teh cotangent for `ϵ`. Is it completely in `dϵ`, or also in the fourth component of `dy`?